source/Nevermore/Advanced/WriteTransaction.cs

```diff
@@ -208,36 +209,51 @@ public class WriteTransaction : ReadTransaction, IRelationalTransaction
             return new DeleteQueryBuilder<TDocument>(ParameterNameGenerator, builder, this);
         }
 
-        public string AllocateId(Type documentType)
+        public TKey AllocateId<TKey>(Type documentType)
         {
             var mapping = configuration.DocumentMaps.Resolve(documentType);
-            return AllocateId(mapping);
+            return AllocateIdForMapping<TKey>(mapping);
         }
 
-        public string AllocateId<TDocument>()
+        public TKey AllocateId<TDocument, TKey>()
         {
             var mapping = configuration.DocumentMaps.Resolve<TDocument>();
-            return AllocateId(mapping);
+            return AllocateIdForMapping<TKey>(mapping);
         }
 
-        string AllocateId(DocumentMap mapping)
+        TKey AllocateIdForMapping<TKey>(DocumentMap mapping)
         {
-            return AllocateId(mapping.TableName, mapping.IdFormat);
+            if (mapping.IdColumn?.Direction == ColumnDirection.FromDatabase)
+                throw new InvalidOperationException($"The document map for {mapping.Type} is configured to use an identity key handler.");
+
+            return (TKey) AllocateIdUsingHandler(mapping);
         }
 
-        public string AllocateId(string tableName, string idPrefix)
+        object AllocateId(DocumentMap mapping)
         {
-            return AllocateId(tableName, key => $"{idPrefix}-{key}");
+            if (mapping.IdColumn?.Direction == ColumnDirection.FromDatabase)
+                throw new InvalidOperationException($"The document map for {mapping.Type} is configured to use an identity key handler.");
+
+            return AllocateIdUsingHandler(mapping);
         }
 
-        string AllocateId(string tableName, Func<int, string> idFormatter)
+        object AllocateIdUsingHandler(DocumentMap mapping)
+        {
+            if (mapping.IdColumn is null || mapping.IsIdentityId)
+                throw new InvalidOperationException($"Cannot allocate an id when an Id column has not been mapped.");
+            return mapping.IdColumn.PrimaryKeyHandler.GetNextKey(keyAllocator, mapping.TableName);
+        }
+
+        public string AllocateId(string tableName, string idPrefix)
```

Comment:
I was wondering if we could get rid of this method and then found this 
![image](https://user-images.githubusercontent.com/347637/123021048-b2427180-d416-11eb-8f53-15a05b233868.png)

Reply:
What do you think about marking this method as obsolete?